File: Design/355. Design Twitter.java
class Tweet {

    private final int tweetId;
    private final int userId;
    private final int time;

    public Tweet(int tweetId, int userId, int time) {
        this.tweetId = tweetId;
        this.userId = userId;
        this.time = time;
    }

    public int getTime() {
        return this.time;
    }

    public int getTweetID() {
        return this.tweetId;
    }

    public int getUserID() {
        return this.userId;
    }

    @Override
    public String toString() {
        return "Time = " + this.time + ", ID = " + this.tweetId + ", user = " + this.userId;
    }
}

class Twitter {

    Map<Integer, LinkedList<Tweet>> tweets;
    Map<Integer, LinkedList<Integer>> followers;
    private int time;

    public Twitter() {
        this.tweets = new HashMap<>();
        this.followers = new HashMap<>();
        this.time = 0;
    }

    public void postTweet(int userId, int tweetId) {
        this.time++;
        Tweet tweet = new Tweet(tweetId, userId, this.time);
        if (!tweets.containsKey(userId)) {
            this.tweets.put(userId, new LinkedList<>(Arrays.asList(tweet)));
        } else {
            this.tweets.get(userId).add(tweet);
        }
    }

    public List<Integer> getNewsFeed(int userId) {
        PriorityQueue<Tweet> tweets_pq = new PriorityQueue<>((t1, t2) -> {
            return t2.getTime() - t1.getTime();
        });
        if (this.tweets.containsKey(userId)) {
            for (Tweet tweet : this.tweets.get(userId)) {
                tweets_pq.add(tweet);
            }
        }
        if (this.followers.containsKey(userId)) {
            for (int followeeId : this.followers.get(userId)) {
                if (this.tweets.containsKey(followeeId)) {
                    for (Tweet tweet : this.tweets.get(followeeId)) {
                        tweets_pq.add(tweet);
                    }
                }
            }
        }
        List<Integer> result = new LinkedList<>();
        while (result.size() < 10 && !tweets_pq.isEmpty()) {
            result.add(tweets_pq.poll().getTweetID());
        }
        return result;
    }

    public void follow(int followerId, int followeeId) {
        if (!followers.containsKey(followerId)) {
            this.followers.put(followerId, new LinkedList<>(Arrays.asList(followeeId)));
        } else if (this.followers.get(followerId).indexOf(followeeId) < 0) {
            this.followers.get(followerId).add(followeeId);
        }
    }

    public void unfollow(int followerId, int followeeId) {
        if (followers.containsKey(followerId)) {
            this.followers.get(followerId).remove(this.followers.get(followerId).indexOf(followeeId));
        }
    }
}

File: Design/355. Design Twitter.js
var Twitter = function () {
    this.tweets = {};
    this.follwers = {};
    this.time = 0;
};

/** 
 * @param {number} userId 
 * @param {number} tweetId
 * @return {void}
 */
Twitter.prototype.postTweet = function (userId, tweetId) {
    this.time++;
    if (!this.tweets.hasOwnProperty(userId)) {
        this.tweets[userId] = [{ tweetId, time: this.time }];
    } else {
        this.tweets[userId].push({ tweetId, time: this.time });
    }
};

/** 
 * @param {number} userId
 * @return {number[]}
List<Integer> getNewsFeed(int userId) Retrieves the 10 most recent tweet IDs in the user's news feed. Each item in the news feed must be posted by users who the user followed or by the user themself
. Tweets must be ordered from most recent to least recent.
*/

Twitter.prototype.getNewsFeed = function (userId) {
    let res = [];
    if (this.tweets.hasOwnProperty(userId)) {
        res = res.concat(this.tweets[userId]);
    }
    if (this.follwers.hasOwnProperty(userId)) {
        this.follwers[userId].forEach(follower => {
            if (this.tweets.hasOwnProperty(follower)) {
                res = res.concat(this.tweets[follower]);
            }
        });
    }
    res.sort((a, b) => b.time - a.time);
    return res.slice(0, 10).map(tweet => tweet.tweetId);
};

/** 
 * @param {number} followerId 
 * @param {number} followeeId
 * @return {void}
 */
Twitter.prototype.follow = function (followerId, followeeId) {
    if (!this.follwers.hasOwnProperty(followerId)) {
        this.follwers[followerId] = [followeeId];
    } else if(!this.follwers[followerId].includes(followeeId)){
        this.follwers[followerId].push(followeeId);
    }
};

/** 
 * @param {number} followerId 
 * @param {number} followeeId
 * @return {void}
 */
Twitter.prototype.unfollow = function (followerId, followeeId) {
    if (this.follwers.hasOwnProperty(followerId)) {
        this.follwers[followerId] = this.follwers[followerId].filter(id => {
            return id !== followeeId;
        });
    }
};